File: App/src/Adaptive.ReactiveTrader.Client.JS/ServiceClients/BlotterServiceClient.ts
 class BlotterServiceClient implements IBlotterServiceClient {
     private _connection: IConnection;
     
     constructor(connection: IConnection) {
         this._connection = connection;
     }

     getTradesStream(): Rx.Observable<TradeDto[]> {
         return this.getTradesStreamFromConnection(this._connection.blotterHubProxy);
     }

     private getTradesStreamFromConnection(blotterHub: HubProxy): Rx.Observable<TradeDto[]> {
         return Rx.Observable.create<TradeDto[]>(observer=> {
             var tradesSubscription = this._connection.allTrades.subscribe(observer);

             console.log("Sending blotter subscription...");
             this._connection.blotterHubProxy.invoke("SubscribeTrades")
                 .done(_=> console.log("Subscribed to blotter"))
                 .fail(ex=> observer.onError(ex));

             var unsubscriptionDisposable = Rx.Disposable.create(()=> {
                 console.log("Sending blotter unsubscription...");

                 this._connection.blotterHubProxy.invoke("UnsubscribeTrades")
                     .done(_=> console.log("Unsubscribed from blotter"))
                     .fail(ex=> console.error("An error occured while unsubscribing from blotter:" + ex));
             });

             return new Rx.CompositeDisposable([tradesSubscription, unsubscriptionDisposable]);
         })
         .publish()
         .refCount();
     }
 }

File: App/src/Adaptive.ReactiveTrader.Client.Domain/ServiceClients/BlotterServiceClient.cs
using System;
using System.Collections.Generic;
using System.Reactive;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain.Transport;
using Adaptive.ReactiveTrader.Shared;
using Adaptive.ReactiveTrader.Shared.DTO.Execution;
using log4net;
using Microsoft.AspNet.SignalR.Client;

namespace Adaptive.ReactiveTrader.Client.Domain.ServiceClients
{
    internal class BlotterServiceClient : ServiceClientBase, IBlotterServiceClient
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(BlotterServiceClient));

        public BlotterServiceClient(IConnectionProvider connectionProvider) : base(connectionProvider)
        {
        }

        public IObservable<IEnumerable<TradeDto>> GetTradesStream()
        {
            return GetResilientStream(connection => GetTradesForConnection(connection.BlotterHubProxy), TimeSpan.FromSeconds(5));
        }

        private static IObservable<IEnumerable<TradeDto>> GetTradesForConnection(IHubProxy blotterHubProxy)
        {
            return Observable.Create<IEnumerable<TradeDto>>(observer =>
            {
                // subscribe to trade feed first, otherwise there is a race condition 
                var spotTradeSubscription = blotterHubProxy.On<IEnumerable<TradeDto>>(ServiceConstants.Client.OnNewTrade, observer.OnNext);

                Log.Info("Sending blotter subscription...");
                var sendSubscriptionDisposable = SendSubscription(blotterHubProxy)
                    .Subscribe(
                        _ => Log.InfoFormat("Subscribed to blotter."),
                        observer.OnError);

                var unsubscriptionDisposable = Disposable.Create(() =>
                {
                    // send unsubscription when the observable gets disposed
                    Log.Info("Sending trades unsubscription...");
                    SendUnsubscription(blotterHubProxy)
                        .Subscribe(
                            _ => Log.InfoFormat("Subscribed to blotter."),
                            ex => Log.WarnFormat("An error occured while unsubscribing to blotter: {0}", ex.Message));
                });
                return new CompositeDisposable { spotTradeSubscription, unsubscriptionDisposable, sendSubscriptionDisposable };
            })
                .Publish()
                .RefCount();
        }

        private static IObservable<Unit> SendSubscription(IHubProxy blotterHubProxy)
        {
            return Observable.FromAsync(() => blotterHubProxy.Invoke(ServiceConstants.Server.SubscribeTrades));
        }

        private static IObservable<Unit> SendUnsubscription(IHubProxy blotterHubProxy)
        {
            return Observable.FromAsync(() => blotterHubProxy.Invoke(ServiceConstants.Server.UnsubscribeTrades));
        } 
    }
}